packages/@react-spectrum/table/src/TableView.tsx

```diff
@@ -442,22 +442,63 @@ function TableVirtualizer({layout, collection, focusedKey, renderView, renderWra
     transitionDuration
   });
 
+  let scrollToItem = useCallback((key) => {
+    let item = collection.getItem(key);
+    let column = collection.columns[0];
+    let virtualizer = state.virtualizer;
+
+    // If the key to be scrolled into view is a column header, we will need to make sure to sync its scroll position
+    // with the body's scroll position after virtualizer.scrollToItem is called so that calling scrollIntoViewport can
+    // properly decide if the column is outside the viewport or not. If we don't, then the column scroll position is handled
+    // in onScroll which is always after our scrollIntoViewport call here.
+    let focusedKeyRect = virtualizer.getView(focusedKey)?.layoutInfo?.rect;
+    let visibleRect = virtualizer.visibleRect;
+    let modality = getInteractionModality();
+    let syncScrollPos;
+
+    if (focusedKeyRect && item?.type === 'column') {
+      // If column can't fit fully in the visible rect, then it will be scrolled into view if the top left (start) corner is
+      // out of view
+      if (focusedKeyRect.width > visibleRect.width || focusedKeyRect.height > visibleRect.height) {
+        if (
+          focusedKeyRect.x < visibleRect.x ||
+          focusedKeyRect.y < visibleRect.y
+        ) {
+          syncScrollPos = true;
+        }
+      } else if (
+        focusedKeyRect.x < visibleRect.x ||
+        focusedKeyRect.maxX > visibleRect.maxX ||
+        focusedKeyRect.y < visibleRect.y ||
+        focusedKeyRect.maxY > visibleRect.maxY
+      ) {
+        syncScrollPos = true;
+      }
+    }
```

Comment:
Won't `scrollIntoViewport` handle scrolling the cell into view if needed? Maybe I'm not understanding what this extra logic is for...

Reply:
This is solving for a similar issue as described in my [comment here](https://github.com/adobe/react-spectrum/pull/3899#discussion_r1088508307). When 
```
virtualizer.scrollToItem(key, {
      duration: 0,
...
```
is called to scroll a table cell into view, the table column header's scroll position isn't actually synced immediately, resulting in something that looks like this:
<img width="1354" alt="image" src="https://user-images.githubusercontent.com/9661127/214993571-04654413-357c-4de2-8800-99609f3e8d8b.png">

In the image above, I moved focus from "Column 5" to "Column 6" (which is slightly out of view) and you can see that the body of the table has scrolled to reflect the new position. The table column header row only updates its scroll position via https://github.com/adobe/react-spectrum/blob/0d06149c05549653cd27feba22dff6bbf4ac8cdc/packages/%40react-spectrum/table/src/TableView.tsx#L507-L520, with `onScroll` occurring before the useEffect and being the call that syncs the header row's position with the body. However, that means we trigger `scrollIntoViewport` before fully repositioning the table column header row scroll position, resulting in https://github.com/adobe/react-spectrum/blob/0d06149c05549653cd27feba22dff6bbf4ac8cdc/packages/%40react-aria/utils/src/scrollIntoView.ts#L120-L122 being triggered and shifting the table unnecessarily. To solve this, I simply update the scroll position of the header row early, and only if I determine that the column would have been scrolled into view already.

I tried previously moving my scrollIntoViewport call out into a useLayoutEffect/useEffect that would trigger on `focusedKey` change instead but that ran into all kinds of issues where it was too early (both were before the `onScroll` call that would update the column header's scroll position)

Reply:
That part I get. Why is syncing the header scroll position conditional though? Does it hurt to always keep them in sync?

Reply:
Ah I see. I had initially wanted to limit the scope of the change but it doesn't seem to have any problematic effects, pushed